src/main/java/plugins/UPnP2/ServiceManager.java

```diff
@@ -0,0 +1,501 @@
+            return -1;
+        }
+
+        IGDRates rates = getRates();
+
+
+        if (rates == null) {
+            return -1;
+        }
+
+        Logger.normal(this, "Upstream MaxBitRate: " + rates.getUpstreamMax());
+
+        return rates.getUpstreamMax();
+    }
+
+    public int getDownstreamMaxBitRate() {
+        Logger.normal(this, "Calling getDownstreamMaxBitRate()");
+
+        waitForBooting();
+
+        if (connectionServices.size() < 0 && commonServices.size() < 0) {
+            return -1;
+        }
+
+        IGDRates rates = getRates();
+
+        if (rates == null) {
+            return -1;
+        }
+
+        Logger.normal(this, "Downstream MaxBitRate: " + rates.getDownstreamMax());
+
+        return rates.getDownstreamMax();
+    }
+
+
+    private IGDRates getRates() {
+
+        final List<Integer> upRates = new ArrayList<>();
+        final List<Integer> downRates = new ArrayList<>();
+
+        for (final Service service : connectionServices) {
+            if (logMINOR) Logger.minor(this, "Service Type: " + service.getServiceType().getType());
+            if (service.getServiceType().getType().equals("WANPPPConnection")
+                    // Make sure the device isn't double natted
+                    // Double natted devices won't have a valid external IP
+                    && detectedIPs.containsKey(service.getDevice().getRoot())
+                    ) {
+
+                new GetLinkLayerMaxBitRates(service, upnpService.getControlPoint()) {
+                    @Override
+                    protected void success(int newUpstreamMaxBitRate, int newDownstreamMaxBitRate) {
+                        if (logMINOR)
+                            Logger.minor(this, "newUpstreamMaxBitRate: " + newUpstreamMaxBitRate);
+                        if (logMINOR)
+                            Logger.minor(this, "newDownstreamMaxBitRate: " +
+                                    newDownstreamMaxBitRate);
+
+                        upRates.add(newUpstreamMaxBitRate);
+                        downRates.add(newDownstreamMaxBitRate);
+                    }
+
+                    @Override
+                    public void failure(ActionInvocation invocation, UpnpResponse operation,
+                                        String defaultMsg) {
+                        Logger.warning(this, "Unable to get MaxBitRates. Reason: " +
+                                defaultMsg);
+                    }
+                }.run(); // Synchronous!
+
+            }
+        }
+
+        if (upRates.size() > 0) {
+            int upRatesSum = 0;
+            for (int rate : upRates) {
+                upRatesSum += rate;
+            }
+            int downRatesSum = 0;
+            for (int rate : downRates) {
+                downRatesSum += rate;
+            }
+            return new IGDRates(upRatesSum, downRatesSum);
+        }
+
+        // We get nothing from GetLinkLayerMaxBitRates. Try GetCommonLinkProperties
```

Comment:
This is almost a duplicate of the first half of this method. It should be possible to extract the logic so it’s in a single place.